lib/internal/streams/from.js

```diff
@@ -34,24 +34,52 @@ function from(Readable, iterable, opts) {
     objectMode: true,
     ...opts
   });
+
   // Reading boolean to protect against _read
   // being called before last iteration completion.
   let reading = false;
+
+  // needToClose boolean if iterator needs to be explicitely closed
+  let needToClose = false;
+
   readable._read = function() {
     if (!reading) {
       reading = true;
       next();
     }
   };
+
+  readable._destroy = async function(error, cb) {
+    try {
+      if (needToClose) {
+        needToClose = false;
+        if (typeof iterator.return === 'function') {
+          const { value } = await iterator.return();
+          await value;
+        }
+      }
+    } catch (e) {
+      if (!error) { // Do not hide current error in case if .return() throws
+        error = e;
+      }
+    } finally {
+      cb(error);
+    }
+  };
+
   async function next() {
     try {
+      needToClose = false;
       const { value, done } = await iterator.next();
```

Comment:
We might need to check `readable.destroyed` after every `await` here. To handle the case where `readable.destroy()` is called before `await` completed...

Reply:
added check